waspc/src/Analyzer/Evaluator.hs

```diff
@@ -1,15 +1,50 @@
+{-# LANGUAGE LambdaCase #-}
+
 module Analyzer.Evaluator
-  ( EvaluationError,
+  ( EvaluationError (..),
     evaluate,
+    Decl,
+    takeDecls,
+    module Analyzer.Evaluator.Types,
   )
 where
 
-import Analyzer.Decl (Decl)
-import Analyzer.TypeChecker (TypedAST)
-import Analyzer.TypeDefinitions (TypeDefinitions)
+import Analyzer.Evaluator.Decl (Decl, takeDecls)
+import Analyzer.Evaluator.EvaluationError
+import Analyzer.Evaluator.Types
+import Analyzer.Type
+import qualified Analyzer.TypeChecker.AST as AST
+import qualified Analyzer.TypeDefinitions as TD
+import Control.Monad.Except
+import Control.Monad.Reader
+import Control.Monad.State
+import qualified Data.HashMap.Strict as H
+import Data.Maybe (fromMaybe)
 
-data EvaluationError
+type Eval a = StateT (H.HashMap String Decl) (ReaderT TD.TypeDefinitions (Except EvaluationError)) a
 
 -- | Evaluate type-checked AST to produce a list of declarations.
-evaluate :: TypeDefinitions -> TypedAST -> Either EvaluationError [Decl]
-evaluate _ _ = Right []
+evaluate :: TD.TypeDefinitions -> AST.TypedAST -> Either EvaluationError [Decl]
+evaluate typeDefs (AST.TypedAST stmts) = runExcept $ flip runReaderT typeDefs $ evalStateT (evalStmts stmts) H.empty
+
+-- TODO:
+-- Currently, trying to reference declarations declared after the current one
+-- fails. There are some solutions mentioned in docs/wasplang that should be
+-- investigated.
+evalStmts :: [AST.TypedStmt] -> Eval [Decl]
+evalStmts = foldr (\stmt -> (<*>) ((:) <$> evalStmt stmt)) (pure [])
+
+evalStmt :: AST.TypedStmt -> Eval Decl
+evalStmt (AST.Decl name param (DeclType declTypeName)) = do
```

Comment:
Uff I thought DeclType here is form Evaluator, it took me some time to figure out it is from Analyzer.Type.
Maybe we should import it qualified and use it like that, qualified, so it is obvious where is it from?

Reply:
Done